sklearn/utils/_param_validation.py

```diff
@@ -0,0 +1,391 @@
+
+    def _mark_if_deprecated(self, option):
+        """Add a deprecated mark to an option if needed."""
+        option_str = f"{option!r}"
+        if option in self.deprecated:
+            option_str = f"{option_str} (deprecated)"
+        return option_str
+
+    def __repr__(self):
+        options_str = (
+            f"{', '.join([self._mark_if_deprecated(o) for o in self.options])}"
+        )
+        return f"a str among {{{options_str}}}"
+
+
+class Interval(_Constraint):
+    """Constraint representing an typed interval.
+
+    Parameters
+    ----------
+    type : {numbers.Integral, numbers.Real}
+        The set of numbers in which to set the interval.
+
+    left : float or int or None
+        The left bound of the interval. None means left bound is -∞.
+
+    right : float, int or None
+        The right bound of the interval. None means right bound is +∞.
+
+    closed : {"left", "right", "both", "neither"}, default="left"
+        Whether the interval is open or closed. Possible choices are:
+
+        - `"left"`: the interval is closed on the left and open on the rigth.
+          It is equivalent to the interval `[ left, right )`.
+        - `"right"`: the interval is closed on the right and open on the left.
+          It is equivalent to the interval `( left, right ]`.
+        - `"both"`: the interval is closed.
+          It is equivalent to the interval `[ left, right ]`.
+        - `"neither"`: the interval is open.
+          It is equivalent to the interval `( left, right )`.
+
+    Notes
+    -----
+    Setting a bound to `None` and setting the interval closed is valid. For instance,
+    strictly speaking, `Interval(0, None, closed="both")` corresponds to
+    `[0, +∞) U {+∞}`.
+    """
+
+    def _validate_params(self):
+        if self.type not in (Integral, Real):
+            raise ValueError(
+                "type must be numbers.Integral or numbers.Real. "
+                f"Got {self.type} instead."
+            )
+
+        if (
+            self.type is Integral
+            and self.left is not None
+            and not isinstance(self.left, Integral)
+        ):
+            raise TypeError(
+                "Expecting left to be an int for an interval over the integers"
+            )
+        if (
+            self.type is Integral
+            and self.left is not None
+            and not isinstance(self.left, Integral)
+        ):
+            raise TypeError(
+                "Expecting left to be an int for an interval over the integers"
+            )
+
+        if self.left is None and self.right is None:
+            raise ValueError(
+                "This interval has no bounds. Set a type constraint only instead."
+            )
+
+    def __init__(self, type, left, right, *, closed="left"):
+        self.type = type
+        self.left = left
+        self.right = right
+        self.closed = closed
+
+        self._validate_params()
```

Comment:
Strange loopy joke: can we specify `_Constraint`s for `_Constraint`s? :upside_down_face:

Reply:
I tried to use this api to validate the parameters of this module but I was not successful. 
Obviously you can't use Interval to validate in the init of Interval otherwise you end up in an infinite loop. But even when trying to avoid these infinite loops I had weird errors that I couldn't solve so I gave up :)

Reply:
Actually I made it work for some of them :)

Reply:
Initially, this just was a pun, not an actual suggestion; but if you've managed to do so, then I cannot help myself from approving it: that's nice! :smiley: